This is a first pass at merging bsuite branch and develop.

Jodasue opened the pull request:
I am creating a this pull request so we can all go over this new code.  I have walked through the changes that were on the Bsuite branch try doing a merge with develop.  Some things I did:
- I updated where bsuite results are being stored
- LSTM was not working and still is having issues (will write more a little lower)
- I took out some some things that I want to be in a separate commit (i.e. a new workflow) 

Some problems:
The async learner doesn't seem to work with bsuite.  I am not sure why, but we need to do something about the workflows asap.  The sync learner seems to be working with bsuite fine.

The LSTM model I think is wrong.  We had to fix the input layers of MLP and LSTM as they were not able to handle different types of gym spaces.  In the branch with unit testing I have dedicated tests that show this flaw.  In looking deeper I think that the way we need to use an LSTM will require more changes than just changing the model build.  I am using the following links to understand what to do:

https://machinelearningmastery.com/reshape-input-data-long-short-term-memory-networks-keras/
https://github.com/MohammadFneish7/Keras_LSTM_Diagram
https://stackoverflow.com/questions/3751900/create-file-path-from-variables
https://github.com/keras-team/keras/issues/7403

I think the issue is that when you use an LSTM you are suppose to pass in a series of experiences not just individual experiences.  This means when we choose LSTM it should also change the experience selection (which now seems to be pushed to the replay buffer).  As far as I know I have not seen this in our current code.  In all fairness I may not understand and could use the help of our AI/ML experts to clarify.

Ultimately, what I am looking for is for some additional eyes on this merge.
- Ted, please look over the changes and see if they make sense to you
- Jamal, we have added the appropriate config file and Ted also added a bsuite.rst file.  Do we need to rebuild the docs?
- Vinay, we need to look into the LSTM and workflow issues.  We should coordinate.

Open question:
Which AI experts should we have double checking our LSTM? Ai, Ted, Himanshu, Tom?

jmohdyusof commented:
bsuite.rst should be in the 'usage' subdirectory with all the other hand-written docs. The index.rst should be updated to include it (again, see the example of all the other docs included in index.rst)

rvinaybharadwaj commented:
The CI tests are failing, please fix them.

rvinaybharadwaj commented:
Add bsuite and dependencies (if any) to setup/test-requirements.txt.

rvinaybharadwaj commented:
I tried running bsuite_driver_all but it is taking a long time, so I tested a few bsuite_ids with the Bsuite env on Darwin and they produce csv logs.

Jodasue commented on exarl/agents/agent_vault/dqn.py:
```diff
@@ -283,8 +294,10 @@ def action(self, state):
         action, policy = self.get_action(state)
-        if not self.is_discrete:
-            action = [self.actions[action]]
+
```
Fixed

Jodasue commented on bsuite/SingleRun.pm:
```diff
@@ -0,0 +1,168 @@
+use strict;
+use Time::HiRes qw(usleep);
+
+my $user = "suet688";
```
Fixed

Jodasue commented:
> The CI tests are failing, please fix them.

Fixed flake8, docs are broken from power grid, and not bothering to pass old tests that we will remove in next merge (we removed set_learner from the api anyway).



> Add bsuite and dependencies (if any) to setup/test-requirements.txt.

I don't think there is any to add at this time.


> I tried running bsuite_driver_all but it is taking a long time, so I tested a few bsuite_ids with the Bsuite env on Darwin and they produce csv logs.

Cleaned the bsuite_driver up and was able to run it on sync learner with limited episodes.  See new comment in bsuite_driver.
